Raygun4UWP/RaygunClient.cs

```diff
@@ -0,0 +1,455 @@
+    {
+      if (ValidateApiKey())
+      {
+        bool canSend = OnSendingCrashReport(originalException, raygunCrashReport);
+        if (canSend)
+        {
+          try
+          {
+            JsonSerializerSettings settings = new JsonSerializerSettings
+            {
+              NullValueHandling = NullValueHandling.Ignore,
+              Formatting = Formatting.None
+            };
+
+            string payload = JsonConvert.SerializeObject(raygunCrashReport, settings);
+
+            if (InternetAvailable())
+            {
+              await SendCrashReport(payload, true);
+            }
+            else
+            {
+              await SaveCrashReport(payload);
+            }
+          }
+          catch (Exception ex)
+          {
+            Debug.WriteLine($"Error Logging Exception to Raygun: {ex.Message}");
+          }
+        }
+      }
+    }
+
+    private bool ValidateApiKey()
+    {
+      if (string.IsNullOrEmpty(Settings.ApiKey))
+      {
+        System.Diagnostics.Debug.WriteLine("ApiKey has not been provided, exception will not be logged");
+        return false;
+      }
+
+      return true;
+    }
+
+    // Returns true if the crash report can be sent, false if the sending is canceled.
+    private bool OnSendingCrashReport(Exception originalException, RaygunCrashReport raygunCrashReport)
+    {
+      bool result = true;
+
+      if (!_handlingRecursiveErrorSending)
+      {
+        EventHandler<RaygunSendingCrashReportEventArgs> handler = SendingCrashReport;
+        if (handler != null)
+        {
+          RaygunSendingCrashReportEventArgs args = new RaygunSendingCrashReportEventArgs(originalException, raygunCrashReport);
+          try
+          {
+            handler(this, args);
+          }
+          catch (Exception e)
+          {
+            // Catch and send exceptions that occur in the SendingCrashReport event handler.
+            // Set the _handlingRecursiveErrorSending flag to prevent infinite errors.
+            _handlingRecursiveErrorSending = true;
+            Send(e);
+            _handlingRecursiveErrorSending = false;
+          }
+
+          result = !args.Cancel;
+        }
+      }
+
+      return result;
+    }
+
+    private async void BeginSendStoredCrashReports()
+    {
+      await SendStoredCrashReports();
+    }
+
+    private async Task SendStoredCrashReports()
+    {
+      if (InternetAvailable())
+      {
+        try
+        {
+          var tempFolder = ApplicationData.Current.TemporaryFolder;
+
+          var raygunFolder = await tempFolder.CreateFolderAsync(OFFLINE_DATA_FOLDER, CreationCollisionOption.OpenIfExists).AsTask().ConfigureAwait(false);
+
+          var files = await raygunFolder.GetFilesAsync().AsTask().ConfigureAwait(false);
+
+          foreach (var file in files)
+          {
+            string text = await FileIO.ReadTextAsync(file).AsTask().ConfigureAwait(false);
+            await SendCrashReport(text, false).ConfigureAwait(false);

```

Comment:
We could probably use finer grained exception handling here so if read on a file fails, we ignore or remove the file and continue with other files. Otherwise this could fail every time it starts and never clear the contents.

Reply:
Added a try/catch block here